dll/kernel32.cpp

```diff
@@ -198,32 +202,106 @@ namespace kernel32 {
 		exit(uExitCode);
 	}
 
-	int WIN_FUNC CreateProcessA(
-		const char *lpApplicationName,
-		char *lpCommandLine,
-		void *lpProcessAttributes,
-		void *lpThreadAttributes,
-		int bInheritHandles,
-		int dwCreationFlags,
-		void *lpEnvironment,
-		const char *lpCurrentDirectory,
-		void *lpStartupInfo,
-		void *lpProcessInformation
-	) {
-		printf("CreateProcessA %s \"%s\" %p %p %d 0x%x %p %s %p %p\n",
-			lpApplicationName,
-			lpCommandLine,
-			lpProcessAttributes,
+	BOOL WIN_FUNC GetExitCodeProcess(HANDLE hProcess, LPDWORD lpExitCode) {
+		DEBUG_LOG("GetExitCodeProcess\n");
+
+		processes::Process* process = processes::processFromHandle(hProcess, false);
+		*lpExitCode = process->exitCode;
+		return 1; // success in retrieval
+	}
+
+	struct PROCESS_INFORMATION {
+		HANDLE hProcess;
+		HANDLE hThread;
+		DWORD  dwProcessId;
+		DWORD  dwThreadId;
+	};
+
+
+ 	BOOL WIN_FUNC CreateProcessA(
+		LPCSTR lpApplicationName,
+		LPSTR lpCommandLine,
+ 		void *lpProcessAttributes,
+ 		void *lpThreadAttributes,
+		BOOL bInheritHandles,
+		DWORD dwCreationFlags,
+		LPVOID lpEnvironment,
+		LPCSTR lpCurrentDirectory,
+ 		void *lpStartupInfo,
+		PROCESS_INFORMATION *lpProcessInformation
+ 	) {
+		DEBUG_LOG("CreateProcessA %s \"%s\" %p %p %d 0x%x %p %s %p %p\n",
+ 			lpApplicationName,
+ 			lpCommandLine,
+ 			lpProcessAttributes,
 			lpThreadAttributes,
 			bInheritHandles,
 			dwCreationFlags,
 			lpEnvironment,
-			lpCurrentDirectory ? lpCurrentDirectory : "<none>",
-			lpStartupInfo,
-			lpProcessInformation
-		);
-		printf("Cannot handle process creation, aborting\n");
-		exit(1);
+			lpCurrentDirectory,
+ 			lpStartupInfo,
+ 			lpProcessInformation
+ 		);
+
+		// Argument parsing
+		// First: how many arguments do we have?
+		size_t argc = 2;
+
+		for (size_t i = 1; i < strlen(lpCommandLine); i++) {
+			if (isspace(lpCommandLine[i]) && !isspace(lpCommandLine[i - 1]))
+				argc++;
+		}
```

Comment:
Should this use the same command line parsing algorithm as we use on startup?

Reply:
I'd definitely love to be reinventing the wheel less here. What exactly did you have in mind? Copying over the path translation and backslash escaping?

Reply:
Oh, hm, I misremembered the direction of that logic, we're doing the opposite here. Never mind!